# Using PEFT with custom models

`peft` allows us to fine-tune models efficiently with LoRA. In this short notebook, we will demonstrate how to train a simple multilayer perceptron (MLP) using `peft`.

## Imports

Make sure that you have the latest version of `peft` installed. To ensure that, run this in your Python environment:
    
    python -m pip install --upgrade peft

In [1]:
import copy
import os

# ignore bnb warnings
os.environ["BITSANDBYTES_NOWELCOME"] = "1"

In [2]:
import peft
import torch
from torch import nn
import torch.nn.functional as F

In [3]:
torch.manual_seed(0)

## Data

We will create a toy dataset consisting of random data for a classification task. There is a little bit of signal in the data, so we should expect that the loss of the model can improve during training.

In [4]:
X = torch.rand((1000, 20))
y = (X.sum(1) > 10).long()

In [5]:
n_train = 800
batch_size = 64

In [6]:
train_dataloader = torch.utils.data.DataLoader(
    torch.utils.data.TensorDataset(X[:n_train], y[:n_train]),
    batch_size=batch_size,
    shuffle=True,
)
eval_dataloader = torch.utils.data.DataLoader(
    torch.utils.data.TensorDataset(X[n_train:], y[n_train:]),
    batch_size=batch_size,
)

## Model

As a model, we use a simple multilayer perceptron (MLP). For demonstration purposes, we use a very large number of hidden units. This is totally overkill for this task but it helps to demonstrate the advantages of `peft`. In more realistic settings, models will also be quite large on average, so this is not far-fetched.

In [7]:
class MLP(nn.Module):
    def __init__(self, num_units_hidden=2000):
        super().__init__()
        self.seq = nn.Sequential(
            nn.Linear(20, num_units_hidden),
            nn.ReLU(),
            nn.Linear(num_units_hidden, num_units_hidden),
            nn.ReLU(),
            nn.Linear(num_units_hidden, 2),
            nn.LogSoftmax(dim=-1),
        )

    def forward(self, X):
        return self.seq(X)

## Training

Here are just a few training hyper-parameters and a simple function that performs the training and evaluation loop.

In [8]:
lr = 0.002
batch_size = 64
max_epochs = 30
device = "cpu" if not torch.cuda.is_available() else "cuda"

In [9]:
def train(model, optimizer, criterion, train_dataloader, eval_dataloader, epochs):
    for epoch in range(epochs):
        model.train()
        train_loss = 0
        for xb, yb in train_dataloader:
            xb = xb.to(device)
            yb = yb.to(device)
            outputs = model(xb)
            loss = criterion(outputs, yb)
            train_loss += loss.detach().float()
            loss.backward()
            optimizer.step()
            optimizer.zero_grad()

        model.eval()
        eval_loss = 0
        for xb, yb in eval_dataloader:
            xb = xb.to(device)
            yb = yb.to(device)
            with torch.no_grad():
                outputs = model(xb)
            loss = criterion(outputs, yb)
            eval_loss += loss.detach().float()

        eval_loss_total = (eval_loss / len(eval_dataloader)).item()
        train_loss_total = (train_loss / len(train_dataloader)).item()
        print(f"{epoch=:<2}  {train_loss_total=:.4f}  {eval_loss_total=:.4f}")

### Training without peft

Let's start without using `peft` to see what we can expect from the model training.

In [10]:
module = MLP().to(device)
optimizer = torch.optim.Adam(module.parameters(), lr=lr)
criterion = nn.CrossEntropyLoss()

In [11]:
%time train(module, optimizer, criterion, train_dataloader, eval_dataloader, epochs=max_epochs)

epoch=0   train_loss_total=0.7970  eval_loss_total=0.6472
epoch=1   train_loss_total=0.5597  eval_loss_total=0.4898
epoch=2   train_loss_total=0.3696  eval_loss_total=0.3323
epoch=3   train_loss_total=0.2364  eval_loss_total=0.5454
epoch=4   train_loss_total=0.2428  eval_loss_total=0.2843
epoch=5   train_loss_total=0.1251  eval_loss_total=0.2514
epoch=6   train_loss_total=0.0952  eval_loss_total=0.2068
epoch=7   train_loss_total=0.0831  eval_loss_total=0.2395
epoch=8   train_loss_total=0.0655  eval_loss_total=0.2524
epoch=9   train_loss_total=0.0380  eval_loss_total=0.3650
epoch=10  train_loss_total=0.0363  eval_loss_total=0.3495
epoch=11  train_loss_total=0.0231  eval_loss_total=0.2360
epoch=12  train_loss_total=0.0162  eval_loss_total=0.2276
epoch=13  train_loss_total=0.0094  eval_loss_total=0.2716
epoch=14  train_loss_total=0.0065  eval_loss_total=0.2237
epoch=15  train_loss_total=0.0054  eval_loss_total=0.2366
epoch=16  train_loss_total=0.0035  eval_loss_total=0.2673
epoch=17  trai

Okay, so we got an eval loss of ~0.26, which is much better than random.

### Training with peft

Now let's train with `peft`. First we check the names of the modules, so that we can configure `peft` to fine-tune the right modules.

In [12]:
[(n, type(m)) for n, m in MLP().named_modules()]

[('', __main__.MLP),
 ('seq', torch.nn.modules.container.Sequential),
 ('seq.0', torch.nn.modules.linear.Linear),
 ('seq.1', torch.nn.modules.activation.ReLU),
 ('seq.2', torch.nn.modules.linear.Linear),
 ('seq.3', torch.nn.modules.activation.ReLU),
 ('seq.4', torch.nn.modules.linear.Linear),
 ('seq.5', torch.nn.modules.activation.LogSoftmax)]

Next we can define the LoRA config. There is nothing special going on here. We set the LoRA rank to 8 and select the layers `seq.0` and `seq.2` to be used for LoRA fine-tuning. As for `seq.4`, which is the output layer, we set it as `module_to_save`, which means it is also trained but no LoRA is applied.

*Note: Not all layers types can be fine-tuned with LoRA. At the moment, linear layers, embeddings, `Conv2D` and `transformers.pytorch_utils.Conv1D` are supported.

In [13]:
config = peft.LoraConfig(
    r=64,
    target_modules=["seq.0", "seq.2"],
    modules_to_save=["seq.4"],
    use_vera=True,
    use_dora=False
)

Now let's create the `peft` model by passing our initial MLP, as well as the config we just defined, to `get_peft_model`.

In [14]:
module = MLP().to(device)
module_copy = copy.deepcopy(module)  # we keep a copy of the original model for later
peft_model = peft.get_peft_model(module, config)
optimizer = torch.optim.Adam(peft_model.parameters(), lr=lr)
criterion = nn.CrossEntropyLoss()
peft_model.print_trainable_parameters()

CAI QUAN QUE GI VAY TROI
trainable params: 8,130 || all params: 4,441,412 || trainable%: 0.18304989494331983


Checking the numbers, we see that only ~1% of parameters are actually trained, which is what we like to see.

Now let's start the training:

In [15]:
for name, param in peft_model.base_model.named_parameters():
    # print(f'{name}: {param.requires_grad}')
    if 'lora_A' in name:
        print(param)

Parameter containing:
tensor([[-0.1156, -0.0532, -0.3305,  ...,  0.1521,  0.1794,  0.1326],
        [-0.0618, -0.1780, -0.0906,  ..., -0.0063,  0.0679,  0.1894],
        [ 0.0018,  0.0567,  0.2044,  ..., -0.0404, -0.0684,  0.1131],
        ...,
        [ 0.0603, -0.0134, -0.2106,  ..., -0.1254,  0.0162,  0.0019],
        [ 0.1604, -0.0451,  0.0738,  ...,  0.0126,  0.0249, -0.1390],
        [-0.1087, -0.0437,  0.0736,  ..., -0.0678,  0.0047,  0.0846]],
       device='cuda:0')
Parameter containing:
tensor([[ 0.3180, -0.0895, -0.0617,  ..., -0.0555,  0.1398,  0.2081],
        [-0.2012, -0.0046,  0.0322,  ...,  0.0619,  0.0147, -0.1102],
        [-0.0607,  0.0882, -0.0503,  ...,  0.2067,  0.0824,  0.0657],
        ...,
        [-0.1188,  0.0197, -0.0532,  ..., -0.1105, -0.0157, -0.0524],
        [ 0.0858,  0.1780,  0.1016,  ..., -0.0052,  0.0871, -0.0894],
        [ 0.0623,  0.0618, -0.1124,  ...,  0.0148,  0.0317, -0.0321]],
       device='cuda:0')


In [16]:
%time train(peft_model, optimizer, criterion, train_dataloader, eval_dataloader, epochs=max_epochs)

epoch=0   train_loss_total=0.6873  eval_loss_total=0.6661
epoch=1   train_loss_total=0.6456  eval_loss_total=0.6366
epoch=2   train_loss_total=0.6092  eval_loss_total=0.6227
epoch=3   train_loss_total=0.5732  eval_loss_total=0.5876
epoch=4   train_loss_total=0.5461  eval_loss_total=0.5657
epoch=5   train_loss_total=0.5169  eval_loss_total=0.5388
epoch=6   train_loss_total=0.4906  eval_loss_total=0.5218
epoch=7   train_loss_total=0.4636  eval_loss_total=0.4940
epoch=8   train_loss_total=0.4415  eval_loss_total=0.4722
epoch=9   train_loss_total=0.4197  eval_loss_total=0.4577
epoch=10  train_loss_total=0.4093  eval_loss_total=0.4378
epoch=11  train_loss_total=0.3886  eval_loss_total=0.4257
epoch=12  train_loss_total=0.3718  eval_loss_total=0.4214
epoch=13  train_loss_total=0.3571  eval_loss_total=0.4089
epoch=14  train_loss_total=0.3444  eval_loss_total=0.3932
epoch=15  train_loss_total=0.3241  eval_loss_total=0.3763
epoch=16  train_loss_total=0.3150  eval_loss_total=0.3694
epoch=17  trai

In the end, we see that the eval loss is very similar to the one we saw earlier when we trained without `peft`. This is quite nice to see, given that we are training a much smaller number of parameters.

#### Check which parameters were updated

Finally, just to check that LoRA was applied as expected, we check what original weights were updated what weights stayed the same.

In [17]:
for name, param in peft_model.base_model.named_parameters():
    if "lora" not in name:
        continue

    print(f"New parameter {name:<13} | {param.numel():>5} parameters | updated")

New parameter model.seq.0.lora_A.default.weight |  1280 parameters | updated
New parameter model.seq.0.lora_B.default.weight | 128000 parameters | updated
New parameter model.seq.0.lora_d.default |    64 parameters | updated
New parameter model.seq.0.lora_b.default |  2000 parameters | updated
New parameter model.seq.2.lora_A.default.weight | 128000 parameters | updated
New parameter model.seq.2.lora_B.default.weight | 128000 parameters | updated
New parameter model.seq.2.lora_d.default |    64 parameters | updated
New parameter model.seq.2.lora_b.default |  2000 parameters | updated


In [18]:
for name, param in peft_model.base_model.named_parameters():
    # print(f'{name}: {param.requires_grad}')
    if 'lora_b' in name:
        print(param)

Parameter containing:
tensor([[-0.4799],
        [ 0.5932],
        [-0.0522],
        ...,
        [-0.0836],
        [ 0.2380],
        [-0.0901]], device='cuda:0', requires_grad=True)
Parameter containing:
tensor([[-0.2096],
        [-0.0550],
        [-0.1477],
        ...,
        [ 0.0445],
        [-0.0350],
        [ 0.2261]], device='cuda:0', requires_grad=True)


In [19]:
for layer in peft_model.base_model.children():
    print(type(layer))

<class '__main__.MLP'>


In [20]:
params_before = dict(module_copy.named_parameters())
for name, param in peft_model.base_model.named_parameters():
    if "lora" in name:
        continue

    name_before = (
        name.partition(".")[-1].replace("original_", "").replace("module.", "").replace("modules_to_save.default.", "")
    )
    param_before = params_before[name_before]
    if torch.allclose(param, param_before):
        print(f"Parameter {name_before:<13} | {param.numel():>7} parameters | not updated")
    else:
        print(f"Parameter {name_before:<13} | {param.numel():>7} parameters | updated")

KeyError: 'seq.0.base_layer.weight'

So we can see that apart from the new LoRA weights that were added, only the last layer was updated. Since the LoRA weights and the last layer have comparitively few parameters, this gives us a big boost in efficiency.

## Sharing the model through Hugging Face Hub

### Pushing the model to HF Hub

With the `peft` model, it is also very easy to push a model the Hugging Face Hub. Below, we demonstrate how it works. It is assumed that you have a valid Hugging Face account and are logged in:

In [ ]:
user = "BenjaminB"  # put your user name here
model_name = "peft-lora-with-custom-model"
model_id = f"{user}/{model_name}"

In [ ]:
peft_model.push_to_hub(model_id);

adapter_model.bin:   0%|          | 0.00/211k [00:00<?, ?B/s]

Upload 1 LFS files:   0%|          | 0/1 [00:00<?, ?it/s]

As we can see, the adapter size is only 211 kB.

### Loading the model from HF Hub

Now, it only takes one step to load the model from HF Hub. To do this, we can use `PeftModel.from_pretrained`, passing our base model and the model ID:

In [ ]:
loaded = peft.PeftModel.from_pretrained(module_copy, model_id)
type(loaded)

peft.peft_model.PeftModel

Let's check that the two models produce the same output:

In [ ]:
y_peft = peft_model(X.to(device))
y_loaded = loaded(X.to(device))
torch.allclose(y_peft, y_loaded)

True

### Clean up

Finally, as a clean up step, you may want to delete the repo.

In [ ]:
from huggingface_hub import delete_repo

In [ ]:
delete_repo(model_id)